Project system reference

aL3891 opened the pull request:
This PR is for issue #256. Its not done but i wanted to gather feedback early on, i've mostly focused on getting all the content in from the code. (that is, all the settable properties)

After doing this first pass i think it would be nice to have the actual type for each property in a more formal way instead of just stating it in the text for each item. Also more inline samples are needed.
This is a start though :) 

Get story straight on:
- [ ] commands
- [ ] preserveCompilationContext
- [ ] languageId
- [ ] frameworkAssemblies NuGet vs GAC story
- [ ] wrappedProject
- [ ] imports
- [ ] What the `default` `languageVersion` is in the compilation options

aL3891 committed:
Initial draft of project reference docs

aL3891 commented:
A few questions regarding the content itself:

* The import property, i cant find any reference to it in the cli code, is it still supported?
* wrappedProject, i dont quite understand what this property does
* Same goes for preserveCompilationContext
* A number of the file properties have version called *builtin that are settable in the file, i've included them for now but i suppose they aren't really ment to be used?
 
Another question, alot of properties use globbing, i've added information for that, but not actually defined globbing anywhere. i imagine that would be a separate document. 

cc @cartermp

cartermp commented:
> Another question, alot of properties use globbing, i've added information for that, but not actually defined globbing anywhere. i imagine that would be a separate document.

I wouldn't worry about making it a separate document.  We can either link to a definition somewhere or provide a quick example.

Re: other issues - @davidfowl could you clarify?

cartermp commented on docs/project-model/index.md:
```diff
@@ -0,0 +1,5 @@
+﻿# .Net core project system
+* [🔧 Overview](overview.md)
+* [🔧 Diffrent scenarios]
```
I'd rename this to "Common Configuration Scenarios".

cartermp commented on docs/project-model/global-json-reference.md:
```diff
@@ -0,0 +1,24 @@
+
+<a name="projects"></a>
+## projects
+An array of string that specifies what folders the build system should search for projects before falling back to resolving dependencies via nuget
```
I don't think this quite captures what the purpose is here.  For example, if I have this:

```json
   "dependencies":{
      "MyDependencyClassLibrary":""
   }
```
There's nothing for NuGet to resolve.  I'd stick to just stating that this is for the build system to understand where you've placed projects you depend on, should you choose to organize things in a way that differs from the (hopefully eventualy?) convention that doesn't require any configuration.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## name
+Type: String
+
+The name of the project, this is also the name of the (nuget) package the project will produce.
```
This also affects the name of the assembly when built.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## version
+Type: String
+
+The semver version of the project, also used for the nuget package.
```
You'll have to either point to a definition of semver (such as a glossary that should be coming in here), or define it yourself.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## description
+Type: String
+
+A longer decsiption of the project.
```
"description"

We'll need to explain if this description serves any purpose other than being a description field in the `project.json` file itself.  E.g., does it affect the description of a NuGet package?

aL3891 replied:
I'm actually not sure where these end up, but i'd guess they are used for NuGet. who would know? (or where abouts would i check in the code?)

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## summary
+Type: String
+
+A short description of the project.
```
Feels weird specifying this *after* the longer description of the project.  Same feedback as above, we'll have to specify what this does outside of just existing in the file.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## copyright
+Type: String
+
+The copyright information for the project.
```
Same feedback as above with respect to usage outside of the `project.json` file.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## title
+Type: String
+
+The friendly name of the project, can contain spaces and special characters not allowed in the name.
```
> can contain spaces and special characters not allowed in the name.

When you say "name", you mean the `name` attribute, yes?

We'll also have to get info on what uses this.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## owners
+Type: String[]
+
+An array of strings with the names of the owners.
```
Owners of what?

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## tags
+Type: String[]
+
+An array of strings with tags for the project, used for searching in nuget.
```
"NuGet"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## testRunner
+Type: String
+
+The name of the testrunner to use with this project, setting this also marks the project as a test project.
```
We'll have to point to a definition for testrunner.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## compilerName
+Type: String
+
+The name of the compiler used for this project, for c# this will be "csc" for example.
```
Default behavior should be specified here.  We should also specify the strings that represent valid compilers.  Currently that's "csc" and "fsc".  We'll have to define what those are as well.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## entryPoint
+Type: String
+
+The entrypoint method for the project, typically Main().
```
We'll have to state default behavior here.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## releaseNotes
+Type: String
+
+Release notes for the project.
```
As with others, we'll have to state what grabs this info and where it goes.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## requireLicenseAcceptance
+Type: Boolean
+
+A boolean that causes a prompt to accept the package license when installing the package to be shown.
```
We'll have to specify that this is *only* relevant to projects that will be NuGet packages.

aL3891 replied:
Aren't all projects NuGet packages now though?

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## embedInteropTypes
+Type: Boolean
+
+If set to true, embeds com interop types in the assembly. 
```
"COM"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## compile
+Type: Globbing String
+
+A globbing string that specifies what files are included in compilation.
```
We'll have to specify an example of why you'd use this.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## content
+Type: Globbing String
+
+A globbing string that specifies what files are considerd content by the build.
```
"considered"

What is "content" in this context?

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## resource
+Type: Globbing String
+
+A globbing string that specifies what files are resource files.
```
I'd change it to "specifies resource files for the project".

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## namedResource
+Type: Globbing String
+
+an object whos keys are names of named resources and whos values are the paths to each resource.
```
"An" and "whose"

I'm confused about the description versus the Type here.

aL3891 replied:
Ah.. copy paste :)

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## contentBuiltIn
+Type: Globbing String
+
+A globbing string that specifies what files are considerd content by the build.
```
As above, what is "content" in this context?

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## tools
+Type: Object
+
+An object that defines package dependencies that are used as tools for the current project, not as references. Packages defined here are available in scripts that run during the build process, but they are not accessible to the code in the project itself. Tools can for exampel include code generators or post build tools that perform tasks related to packing.
```
"example" and "post-build"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## commands
+Type: Object
+
+Commands are depricated in the cli. (?)
```
Heh, unfortunately I can't answer that.  We can leave this as a question mark or mark it with a **TODO** for now.

aL3891 replied:
Yeah.. they are still settable in the code, so i thought i'd leave them in there, but i was at least under the impression that they were going away.. Who might know for sure?

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+### define
+Type: String[]
+
+A list of defines such as "DEBUG" or "TRACE" that can be used in conditional compilation in the code.
```
I'd call out here that this is used per-target in the `frameworks` section for cross-targeting scenarios.  Example [here](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/project.json#L81)

aL3891 replied:
Do you mean like "This can also be used to set target specific defines[...]" or "The target framework also gets its own define automatically"? i'm not quite sure what the behavior is but i think there is already an automatic define for each target.. i'll look into that though

cartermp replied:
I'm referring to the idea that you'd primarily use this in multi-targeting scenarios, where there either isn't a default definition the compiler understands, or you want to use a better name than the default name.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+### emitEntryPoint
+Type: Boolean
+
+When set to true, the project will compile to an exe file. if se tto false, the project will produce a dll.
```
"If set to"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+## frameworks
+Type: Object
+
+An object whose properties specifies what frameworks this project supports, such as .NetFramework or UWP. Must be a valid target framework moniker (TFM). Each value is an object that can contain information specific to this framework such as compilationoptions, analyzeroptions as well as the following properties.
```
"specify",  ".NET Framework", "Universal Windows Platform (UWP)", "Target Framework Moniker", "`compilationOptions`", and "`analyzerOptions`"

I'd explicitly call out "For example:" for this and all other examples.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+### dependencies
+Type: Object
+
+Dependencies that are specific for this framework. Useful if for example one framework has built in support for something that another framework doesnt, or if a diffrent version of a library is required for a perticular framework.
```
I'd change the second sentence to something like:

"This is useful in scenarios where you cannot simply specify a package-level dependency across all targets.  Reasons for this can include one target lacking built-in support that other targets have, or requiring a different version of a dependency than other targets."

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+### frameworkAssemblies
+Type: Object
+
+Similar to dependencies but contains reference to assemblies in the GAC that are not nuget packages. Can also specify the version to use as well as the dependency type.
```
We'll have to point to a definition for GAC

It's also important to call out that this is for .NET Framework targets only.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,494 @@
+### imports
+Type: String
+
+Specifies portable profiles this project is compatible with. (?)
```
This is a **TODO** as well.

cartermp commented:
Left some specific feedback.  Overall, I like the content.  Thanks for your work so far!

Some larger things remain that we'll have to address for the reference doc:

- Defining the default behavior for each
- Specifying which is required and for what types of code (e.g. there's no need to specify `"emitEntryPoint":false` if you're writing a library)
- Specifying examples for each (sounds like you're going to address that already)
- Point any piece of terminology to a glossary (this may end up being a longer-term change)

aL3891 commented:
Thanks for the feedback! (and sorry about the spelling)
Should i add a glossary page as well in this PR or make a new one? I suppose it would be a global one for the docs. (be its own folder)

cartermp commented:
I'd make a new PR.  @SamuelEnglard has been contributing a ton on the documentation in the CoreFX repo and made a great point about bringing over a glossary of terms over here.  It's definitely something we want to do.

aL3891 commented:
@cartermp cool! is the intention to have all the corefx docs in here as well down the line? 

@DamianEdwards you mentioned docs around project.json and the project system in the community standup, i just wanted to make you aware of this PR and ask if those docs will go here or somewhere else?

SamuelEnglard commented:
@aL3891 There won't be a perfect duplication between corefx and here, but some of the information should be shared (see https://github.com/dotnet/corefx/issues/5899 for more details)

cartermp commented:
@aL3891 I don't think we'll have too much of that content over here (at least not for now).  Some stuff like the .NET Platform Standard will have corresponding articles (or even just brought over here), but since a lot of that documentation is tailored specifically to corefx it'll stay there.

The approach we're taking is to come up with a good strategy for a particular area and then see what content we can use from somewhere else.  In this particular case, the glossary of terms from corefx could have some of its content brought over here.

aL3891 commented:
Allright, addressed most of the feedback, however some things like all the places `summary`, `description` and those kinds of fields go could use some more investigation. 

There is also a few properties that i still havent discern hoe the work in the cli, like `preserveCompilationContext`, `imports`, `wrappedProject` and the `bin` property of the frameworks object. i could really use some help with that, @davidfowl / @DamianEdwards if you could take alook or point to someone who knows that'd be great.

aL3891 commented:
Thanks @SamuelEnglard and @cartermp, i'll look at those for the glossary! From a user perspective though, where will they go to find the corefx docs? to http://dotnet.github.io or somewhere else?

aL3891 commented:
@cartermp @blackdwarf Any additional feedback on this one?

cartermp commented:
@aL3891 I'll take a look at your changes today.

Re: glossary, we'll probably have a glossary in the table of contents of the site.  We'll have to work out where that goes.

aL3891 commented:
Thanks! 

I thought of the glossary as another top level article in this repo, but i havent been able to do any work on that unfortunately

cartermp commented:
No worries.  We'll figure something out.  Could probably put up a quick PR to get a small glossary in so we can start linking to it.  @blackdwarf thoughts?

fubar-coder commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,844 @@
+Type: Object
+
+TODO: Investigate actual status
+Commands are depricated in the cli. (?)
```
It's `deprecated`, not `depricated`.

shahid-pk commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,846 @@
+
+TODO
+Specifies other framework profiles profiles that this project is compatible with.
+
```
Nit : profiles repeated

aL3891 commented:
@cartermp Did you have a chance to take a look?

cartermp commented:
@aL3891 not yet, sorry!  I got pulled into other things.  I blocked out an hour today so I can get this reviewed.

aL3891 commented:
Thanks! :)

cartermp commented on docs/project-model/overview.md:
```diff
@@ -0,0 +1,8 @@
+# .Net core project system
+
+Since .net core can be used on multiple platform it cannot rely on a single editor or enviroment. Because of this, and to prevent every editor to make their own implementation, the project system for .net core is now part of the platform.
```
This frames the project model as an evolution of something specific to an editor or environment.  While true (to some extent), I don't think that's what people will be coming here to read about.  I think it would be better to describe what it is and highlight some of the "need to know" highlights of the project model.

In fact, I think we should hold off on this overall conceptual description right now - there's some churn going on internally right now such that what gets written today could end up being patently false a month from now.  We'll need to get the overall story straight before we can have a definitive conceptual doc describing the project model as a whole.

aL3891 replied:
I really ment this section to just be a placeholder, but maybe its better to remove it completely for now until the churn is over?

cartermp replied:
Yeah, probably a good idea.

cartermp commented on docs/project-model/overview.md:
```diff
@@ -0,0 +1,8 @@
+# .Net core project system
```
".NET core" ---> ".NET Core"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+For example:
+
+    {
+        "entryPoint": "ADiffrentMethod"
```
"ADiffrentMethod" ---> "ADifferentMethod"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+## compilerName
+Type: String
+
+The name of the compiler used for this project. `csc` by default. currently `csc`, the c# compiler or `fsc`, the f# compiler are supported.
```
"currently" ---> "Currently"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+## testRunner
+Type: String
+
+The name of the testrunner, such as nuint or xunit, to use with this project, setting this also marks the project as a test project.
```
"NUnit" and "xUnit".

I'd also make that text links to NUnit and xUnit pages that tell you how to test with .NET Core.

aL3891 replied:
Are there such pages in the docs right now? i could add links to the NUnit/xUnit home pages in the mean time

cartermp replied:
I know that xUnit has a page, but now that I think about it it's still DNX-based.  Probably better to link to the home pages for now.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+### nowarn
+Type: String[]
+
+A list of warnings to ignore.
```
What list of warnings?  The example seems to indicate that warnings aren't output when you compile.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+### languageVersion
+Type: String
+
+The version of the language used by the compiler: ISO-1, ISO-2, 3, 4, 5, 6, or Default
```
Is the default the latest version for each language?  At least that's what it seems like to me.  May be worth clarifying.

aL3891 replied:
i dont know for sure.. it seems likely that its the latest one though

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+### xmlDoc
+Type: Boolean
+
+Enables xml docs to be generated from comments in the sourcecode.
```
"sourcecode" ---> "source code"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+## configurations
+Type: Object
+
+An object whos properties define different configurations for this project, such as Debug and Release. Each value is an object that can contain a compilationoptions object with options specific for this coniguration.
```
"compilationoptions" ---> "compilationOptions"

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+
+Specifies what frameworks this project supports, such as .NET Framework or Universal Windows Platform (UWP). Must be a valid Target Framework Moniker (TFM). Each value is an object that can contain information specific to this framework such as `compilationoptions`, `analyzeroptions` as well as the following properties.
+
+For example:
```
This example doesn't show a TFM with specific configurations for that target.

cartermp commented on docs/project-model/project-json-reference.md:
```diff
@@ -0,0 +1,847 @@
+### frameworkAssemblies
+Type: Object
+
+Similar to dependencies but contains reference to assemblies in the GAC that are not nuget packages. Can also specify the version to use as well as the dependency type. Only used by the .NET Framework and Portable Class Library (PCL) references.
```
> contains reference to assemblies in the GAC that are not nuget packages

This is actually something [we're looking to change](https://github.com/dotnet/corefx/issues/3934#issuecomment-179963211).  That effort is blocked right now, so I think it's okay to state that this is for GAC reference assemblies for the time being.

cartermp commented:
Made another pass through.  Good work so far!  I put a list of checks in your initial comment to we can more easily keep track of the areas where there's still TODOs or we're not quite clear on what the story is.

aL3891 commented:
@cartermp I've adressed the feedback but I'm still not sure about the language version thing. I've added it to the checkboxes in the OP

aL3891 commented:
So.. How are we doing? :)

cartermp commented:
Left a few comments, sorry for the delay - things are getting busy over here with //build coming up.

Since there's still plenty of things to get the story straight on (the checkboxes), what do you think about tracking those in an issue instead?  We can then address them through multiple PRs as needed.  Having one PR with tons of comments gets really slow to load.

I'm thinking about merging this and perhaps putting some text in there stating that this isn't complete yet.  I think I'd rather have something that's incomplete but up there than wait a long time before everything gets addressed.

aL3891 commented:
no worries, i understand things are stressed right now :) 
I'll take a look at those comments, but i agree the outstanding point should be issues especially since some of them are quite big/complicated. I can create them if you want, unless you had some specific grouping in mind.

cartermp commented:
LGTM for now - let's get this merged in and I'll track things in an Issue.  Mind rebasing?

aL3891 commented:
Rebased (i hope, im still a bit of a git noob) 
@cartermp

cartermp commented:
@aL3891 doesn't seem like it.  Have you use `$ git rebase -i master`?  That'll bring you into the interactive window where you can choose to "fixup" commits (those basically squash that commit into the previous one). For example:

If I had 3 commits, but the last two were just addressing some feedback in a PR and I wanted to just have 1 commit so there's a clean history, I'd type `$ git rebase -i master` which would open the following in vim:

p commit1 - the real deal
p commit2 - just some feedback
p commti3 - whoops missed something

To "squash" commit2 and commit3 so that commit1 appeared as the only commit, I'd change it to this:

p commit1 - the real deal
f commit2 - just some feedback
f commti3 - whoops missed something

(`f` is short for "fixup")

And the rebase would show only commit1, but with all the changes of the other commits.  I could also amend that commit with `$ git commit --amend`, which will open the editor and let me change the commit message so that I could make it sound nicer if I wanted.

I'd then force push to the branch to update: `$ git push -f`, since that's required to update like this.

Kind of a lot of info, but git's complex :smile:

aL3891 commented:
Ah, sorry i did a rebase on to the core docs master, i figured you'd do a squash when merging sorry :D i'll try again

cartermp commented:
@aL3891 no worries.  The first time I did this I may or may not have completely destroyed multiple things on my machine :smile:

aL3891 commented:
Allright i think i got it @cartermp

cartermp commented:
:+1:
